spec/app/ics-100-atomic-swap/README.md

```diff
@@ -0,0 +1,593 @@
+  counterpartyVersion: string) => (version: string, err: Error) {
+  // only unordered channels allowed
+  abortTransactionUnless(order === UNORDERED)
+  // assert that version is "ics100-1"
+  abortTransactionUnless(counterpartyVersion === "ics100-1")
+  
+  // allocate an escrow address
+  channelEscrowAddresses[channelIdentifier] = newAddress()
+  
+  return "ics100-1", nil
+}
+```
+
+```typescript
+function onChanOpenAck(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  counterpartyVersion: string
+) {
+  // port has already been validated
+  // assert that counterparty selected version is "ics100-1"
+  abortTransactionUnless(counterpartyVersion === "ics100-1");
+}
+```
+
+```typescript
+function onChanOpenConfirm(portIdentifier: Identifier, channelIdentifier: Identifier) {
+  // accept channel confirmations, port has already been validated, version has already been validated
+}
+```
+
+```typescript
+function onChanCloseInit(portIdentifier: Identifier, channelIdentifier: Identifier) {
+  // always abort transaction
+  abortTransactionUnless(FALSE);
+}
+```
+
+```typescript
+function onChanCloseConfirm(portIdentifier: Identifier, channelIdentifier: Identifier) {
+  // no action necessary
+}
+```
+
+#### Packet relay
+
+`sendAtomicSwapPacket` must be called by a transaction handler in the module which performs appropriate signature checks, specific to the account owner on the host state machine.
+
+```typescript
+function sendAtomicSwapPacket(
+    swapPacket AtomicSwapPacketData, 
+    sourcePort, sourceChannel, timeoutHeight, timeoutTimestamp: int64
+) {
+    // send packet using the interface defined in ICS4
+    handler.sendPacket(
+      getCapability("port"),
+      sourcePort,
+      sourceChannel,
+      timeoutHeight,
+      timeoutTimestamp,
+      swapPacket.getBytes(), // Should be proto marshalled bytes.
+    )
+}
+```
+
+`onRecvPacket` is called by the routing module when a packet addressed to this module has been received.
+
+```typescript
+function onRecvPacket(packet channeltypes.Packet) {
+  const swapPaket: AtomicSwapPacketData =  packet.data
+  switch swapPaket.type {
+      case TYPE_MSG_MAKE_SWAP:
+        const makeMsg = protobuf.decode(swapPaket.data)
+        
+        // check if buy_token is native token on the taker chain
+        const supply = bank.getSuppy(makeMsg.buyToken.denom)
+        abortTransactionUnless(supply > 0)
+        
+        // create and save order on the taker chain.
+        const order = OrderBook.createOrder(makeMsg)
+        order.status = Status.SYNC
+        order.portIdOnTakerChain = packet.destinationPort
+        order.channelIdOnTakerChain = packet.destinationChannel
+        //saves order to store
+        store.save(order)
+        break;
+      case TYPE_MSG_TAKE_SWAP:
+        const takeMsg = protobuf.decode(swapPaket.data)
+        const order = store.findOrderById(takeMsg.orderId)
```

Comment:
These should probably be keyed on the port and channel IDs of the respective channel ends to avoid potential collisions

Reply:
added